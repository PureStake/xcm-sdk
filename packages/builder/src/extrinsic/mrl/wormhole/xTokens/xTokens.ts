import {
  convertAddressTo32Bytes,
  getMultilocationDerivedAddresses,
} from '@moonbeam-network/xcm-utils';
import { AssetAmount } from '@moonbeam-network/xcm-types';
import { Address, encodeFunctionData } from 'viem';
import { Chain, Wormhole } from '@wormhole-foundation/sdk-connect';
import { EvmPlatform } from '@wormhole-foundation/sdk-evm';
import { ExtrinsicBuilder } from '../../../ExtrinsicBuilder';
import { MrlExtrinsicConfigBuilder } from '../../../ExtrinsicBuilder.interfaces';
import { ExtrinsicConfig } from '../../../ExtrinsicConfig';
import { BATCH_CONTRACT_ABI } from './BatchContractAbi';
import { ERC20_ABI } from './Erc20ABI';
import { TOKEN_BRIDGE_ABI } from './TokenBridgeAbi';
import { TOKEN_BRIDGE_RELAYER_ABI } from './TokenBrigdeRelayerAbi';

// TODO: Can we move them somewhere?
const BUY_EXECUTION_FEE = 100_000_000_000_000_000n;
const CROSS_CHAIN_FEE = 100_000_000_000_000_000n;
export const BATCH_CONTRACT_ADDRESS =
  '0x0000000000000000000000000000000000000808';

export function xTokens() {
  return {
    transfer: ({
      isAutomatic,
    }: {
      isAutomatic: boolean;
    }): MrlExtrinsicConfigBuilder => ({
      build: (params) => {
        const {
          asset,
          destination,
          destinationAddress,
          moonApi,
          moonAsset,
          moonChain,
          moonGasLimit,
          sourceApi,
        } = params;

        if (!destination.wh?.name) {
          throw new Error('Destination chain does not have a wormhole name');
        }

        const { transfer } = sourceApi.tx.xTokens;
        const builder = ExtrinsicBuilder().xTokens().transfer();

        const assetTx = transfer(builder.build(params).getArgs(transfer));
        /*
         * TODO: We need to move it to AssetRoute and receive it in build params.
         * In the future we could also check if wallet already has necessary DEV/GLMR to pay execution fees on Moonbase/Moonbeam.
         * Also we need to move fees to AssetRoute.
         */
        const executionAssetTx = transfer(
          builder
            .build({
              ...params,
              asset: AssetAmount.fromChainAsset(moonAsset, {
                amount: CROSS_CHAIN_FEE + BUY_EXECUTION_FEE,
              }),
            })
            .getArgs(transfer),
        );

        const wh = new Wormhole(moonChain.isTestChain ? 'Testnet' : 'Mainnet', [
          EvmPlatform,
        ]);
        const whDestinationChain = wh.getChain(destination.wh.name).config
          .chainId;
        const whContractAddress = (
          isAutomatic
            ? wh.getChain('Moonbeam').config.contracts.tokenBridgeRelayer
            : wh.getChain('Moonbeam').config.contracts.tokenBridge
        ) as Address;

        const tokenAddressOnMoonChain = moonChain.getChainAsset(asset)
          .address as Address;
        const tokenAmountOnMoonChain = asset.convertDecimals(
          moonChain.getChainAsset(asset).decimals,
        ).amount;

        const approveTx = encodeFunctionData({
          abi: ERC20_ABI,
          functionName: 'approve',
          args: [whContractAddress, tokenAmountOnMoonChain],
        });
        const transferTx = isAutomatic
          ? encodeFunctionData({
              abi: TOKEN_BRIDGE_RELAYER_ABI,
              functionName: 'transferTokensWithRelay',
              args: [
                tokenAddressOnMoonChain,
                tokenAmountOnMoonChain,
                0,
                whDestinationChain,
                destinationAddress,
                0,
              ],
            })
          : encodeFunctionData({
              abi: TOKEN_BRIDGE_ABI,
              functionName: 'transferTokens',
              args: [
                tokenAddressOnMoonChain,
                tokenAmountOnMoonChain,
                whDestinationChain,
                convertAddressTo32Bytes(destinationAddress) as Address,
                0n,
                0,
              ],
            });
        const batchAll = encodeFunctionData({
          abi: BATCH_CONTRACT_ABI,
          functionName: 'batchAll',
          args: [
            [tokenAddressOnMoonChain, whContractAddress],
            [0n, 0n], // Value to send for each call
            [approveTx, transferTx], // Call data for each call
            [], // Gas limit for each call
          ],
        });

        const { address20 } = getMultilocationDerivedAddresses({
          address: destinationAddress,
          paraId: moonChain.parachainId,
          isParents: true,
        });

        const transact = moonApi.tx.ethereumXcm.transact({
          V2: {
            gasLimit: moonGasLimit,
            action: {
              Call: BATCH_CONTRACT_ADDRESS,
            },
            value: 0,
            input: batchAll,
          },
        });

        const txWeight = (await transact.paymentInfo(address20)).weight;

        const send = sourceApi.tx.polkadotXcm.send(
          {
            V3: {
              parents: 1,
              interior: { X1: { Parachain: moonChain.parachainId } },
            },
          },
          {
            V3: [
              {
                WithdrawAsset: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                ],
              },
              {
                BuyExecution: {
                  fees: {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                  weightLimit: 'Unlimited',
                },
              },
              {
                Transact: {
                  originKind: 'SovereignAccount',
                  requireWeightAtMost: {
                    refTime: txWeight.refTime,
                    proofSize: txWeight.proofSize,
                  },
                  call: {
                    encoded: transact.method.toHex(),
                  },
                },
              },
              {
                SetAppendix: [
                  { RefundSurplus: {} },
                  {
                    DepositAsset: {
                      assets: { Wild: { AllCounted: 1 } },
                      beneficiary: {
                        parents: 0,
                        interior: {
                          X1: { AccountKey20: { key: address20 } },
                        },
                      },
                    },
                  },
                ],
              },
            ],
          },
        );

        return new ExtrinsicConfig({
          module: 'utility',
          func: 'batchAll',
          getArgs: () => [assetTx, executionAssetTx, send],
        });
      },
    }),
  };
}
