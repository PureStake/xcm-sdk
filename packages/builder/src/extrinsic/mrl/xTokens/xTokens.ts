import { getMultilocationDerivedAddresses } from '@moonbeam-network/xcm-utils';
import { AssetAmount } from '@moonbeam-network/xcm-types';
import { ExtrinsicBuilder } from '../../ExtrinsicBuilder';
import { MrlExtrinsicConfigBuilder } from '../../ExtrinsicBuilder.interfaces';
import { ExtrinsicConfig } from '../../ExtrinsicConfig';

const BUY_EXECUTION_FEE = 100_000_000_000_000_000n;
const CROSS_CHAIN_FEE = 100_000_000_000_000_000n;

export function xTokens() {
  return {
    transfer: (): MrlExtrinsicConfigBuilder => ({
      build: (params) => {
        const {
          destinationAddress: address,
          sourceApi: api,
          moonChain,
          moonAsset,
        } = params;
        const { transfer } = api.tx.xTokens;
        const builder = ExtrinsicBuilder().xTokens().transfer();
        const { address20 } = getMultilocationDerivedAddresses({
          address,
          paraId: moonChain.parachainId,
          isParents: true,
        });

        const asset = transfer(builder.build(params).getArgs(transfer));
        /*
         * TODO: We need to move it to AssetRoute and receive it in build params.
         * In the future we could also check if wallet already has necessary DEV/GLMR to pay execution fees on Moonbase/Moonbeam.
         * Also we need to move fees to AssetRoute.
         */
        const executionAsset = transfer(
          builder
            .build({
              ...params,
              asset: AssetAmount.fromChainAsset(moonAsset, {
                amount: CROSS_CHAIN_FEE + BUY_EXECUTION_FEE,
              }),
            })
            .getArgs(transfer),
        );
        const send = api.tx.polkadotXcm.send(
          {
            V3: {
              parents: 1,
              interior: { X1: { Parachain: moonChain.parachainId } },
            },
          },
          {
            V3: [
              {
                WithdrawAsset: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                ],
              },
              {
                BuyExecution: {
                  fees: {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                  weightLimit: 'Unlimited',
                },
              },
              {
                Transact: {
                  originKind: 'SovereignAccount',
                  requireWeightAtMost: {
                    refTime: txWeight.refTime,
                    proofSize: txWeight.proofSize,
                  },
                  call: {
                    encoded: transact.method.toHex(),
                  },
                },
              },
              {
                SetAppendix: [
                  { RefundSurplus: {} },
                  {
                    DepositAsset: {
                      assets: { Wild: { AllCounted: 1 } },
                      beneficiary: {
                        parents: 0,
                        interior: {
                          X1: { AccountKey20: { key: address20 } },
                        },
                      },
                    },
                  },
                ],
              },
            ],
          },
        );

        return new ExtrinsicConfig({
          module: 'utility',
          func: 'batchAll',
          getArgs: () => [asset, executionAsset, send],
        });
      },
    }),
  };
}
