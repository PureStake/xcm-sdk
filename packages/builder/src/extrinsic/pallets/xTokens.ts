/* eslint-disable sort-keys */
import { ExtrinsicConfig } from '../ExtrinsicConfig';
import { ExtrinsicConfigBuilder } from '../ExtrinsicConfigBuilder.interfaces';

export function xTokens() {
  return {
    asset: (): ExtrinsicConfigBuilder => ({
      build: ({ asset }) =>
        new ExtrinsicConfig({
          pallet: 'assets',
          method: 'asset',
          args: [asset],
        }),
    }),
  };
}
