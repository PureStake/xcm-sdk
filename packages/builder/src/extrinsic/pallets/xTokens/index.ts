export * from './xTokens';
export * from './xTokens.interfaces';
export * from './xTokens.utils';
