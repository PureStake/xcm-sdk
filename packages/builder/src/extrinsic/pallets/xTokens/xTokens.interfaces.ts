export type XTokensWeightLimit =
  | number
  | { Limited: number }
  | { Limited: { refTime: number; proofSize: number } };
