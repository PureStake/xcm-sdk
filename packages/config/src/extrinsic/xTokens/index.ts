export * from './xTokens';
export * from './xTokens.constants';
export * from './xTokens.interfaces';
