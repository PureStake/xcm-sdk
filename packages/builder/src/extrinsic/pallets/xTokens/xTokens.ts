/* eslint-disable sort-keys */
import { ExtrinsicConfigBuilder } from '../../ExtrinsicBuilder.interfaces';
import { getExtrinsicArgumentVersion } from '../../ExtrinsicBuilder.utils';
import { ExtrinsicConfig } from '../../ExtrinsicConfig';
import { getWeight } from './xTokens.utils';

const pallet = 'xTokens';

export function xTokens() {
  return {
    transfer: (): ExtrinsicConfigBuilder => ({
      build: ({ address, amount, asset, destination, source }) =>
        new ExtrinsicConfig({
          module: pallet,
          func: 'transfer',
          getArgs: (func) => {
            const version = getExtrinsicArgumentVersion(func, 2);

            return [
              asset,
              amount,
              {
                [version]: {
                  parents: 1,
                  interior: {
                    X2: [
                      {
                        Parachain: destination.parachainId,
                      },
                      {
                        AccountKey20: {
                          network: 'Any',
                          key: address,
                        },
                      },
                    ],
                  },
                },
              },
              getWeight(source.weight, func),
            ];
          },
        }),
    }),
    transferMultiAsset: (
      originParachainId: number,
    ): ExtrinsicConfigBuilder => ({
      build: ({ address, amount, asset, destination, source }) =>
        new ExtrinsicConfig({
          module: pallet,
          func: 'transferMultiAsset',
          getArgs: (func) => [
            {
              V1: {
                id: {
                  Concrete: {
                    parents: 1,
                    interior: {
                      X2: [
                        {
                          Parachain: originParachainId,
                        },
                        {
                          GeneralKey: asset,
                        },
                      ],
                    },
                  },
                },
                fun: {
                  Fungible: amount,
                },
              },
            },
            {
              V1: {
                parents: 1,
                interior: {
                  X2: [
                    {
                      Parachain: destination.parachainId,
                    },
                    {
                      AccountKey20: {
                        network: 'Any',
                        key: address,
                      },
                    },
                  ],
                },
              },
            },
            getWeight(source.weight, func),
          ],
        }),
    }),
    transferMultiCurrencies: (): ExtrinsicConfigBuilder => ({
      build: ({ address, amount, asset, destination, fee, feeAsset, source }) =>
        new ExtrinsicConfig({
          module: pallet,
          func: 'transferMultiCurrencies',
          getArgs: (func) => [
            [
              [asset, amount],
              [feeAsset, fee],
            ],
            1,
            {
              V1: {
                parents: 1,
                interior: {
                  X2: [
                    {
                      Parachain: destination.parachainId,
                    },
                    {
                      AccountKey20: {
                        network: 'Any',
                        key: address,
                      },
                    },
                  ],
                },
              },
            },
            getWeight(source.weight, func),
          ],
        }),
    }),
  };
}
