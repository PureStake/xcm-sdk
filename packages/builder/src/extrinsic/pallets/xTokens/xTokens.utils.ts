import { SubmittableExtrinsicFunction } from '@polkadot/api/types';
import { XTokensWeightLimit } from './xTokens.interfaces';

/**
 * Workaround for weight parameter type change in xTokens pallet
 * https://github.com/open-web3-stack/open-runtime-module-library/pull/841
 */
export function getWeight(
  weight: number,
  func?: SubmittableExtrinsicFunction<'promise'>,
): XTokensWeightLimit {
  const type = func?.meta.args.at(-1)?.type;

  if (!type) {
    return weight;
  }

  if (type.eq('XcmV2WeightLimit')) {
    return {
      Limited: weight,
    };
  }

  if (type.eq('XcmV3WeightLimit')) {
    return {
      Limited: { proofSize: 0, refTime: weight },
    };
  }

  return weight;
}
