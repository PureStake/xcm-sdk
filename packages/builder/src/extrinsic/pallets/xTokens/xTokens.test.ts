import { buildParamsMock } from '../../../../fixtures';
import { xTokens } from './xTokens';

describe('xTokens', () => {
  describe('transfer', () => {
    const extrinsic = xTokens().transfer().build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs()).toMatchSnapshot();
    });
  });

  describe('transferMultiAsset', () => {
    const extrinsic = xTokens().transferMultiAsset().build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs()).toMatchSnapshot();
    });
  });

  describe('transferMultiCurrencies', () => {
    const extrinsic = xTokens()
      .transferMultiCurrencies()
      .build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs()).toMatchSnapshot();
    });
  });
});
