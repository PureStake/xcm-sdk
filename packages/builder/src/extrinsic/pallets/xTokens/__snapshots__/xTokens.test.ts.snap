// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xTokens transfer should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transfer",
  "getArgs": [Function],
  "module": "xTokens",
  "type": "Substrate",
}
`;

exports[`xTokens transfer should get correct arguments 1`] = `
[
  "USDT",
  99000000000n,
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
              "network": "Any",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  1000000000,
]
`;

exports[`xTokens transferMultiAsset x1 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMultiasset",
  "getArgs": [Function],
  "module": "xTokens",
  "type": "Substrate",
}
`;

exports[`xTokens transferMultiAsset x1 should get correct arguments 1`] = `
[
  {
    "V1": {
      "fun": {
        "Fungible": 99000000000n,
      },
      "id": {
        "Concrete": {
          "interior": {
            "X1": {
              "Parachain": 100,
            },
          },
          "parents": 1,
        },
      },
    },
  },
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
              "network": "Any",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;

exports[`xTokens transferMultiAsset x2 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMultiasset",
  "getArgs": [Function],
  "module": "xTokens",
  "type": "Substrate",
}
`;

exports[`xTokens transferMultiAsset x2 should get correct arguments 1`] = `
[
  {
    "V1": {
      "fun": {
        "Fungible": 99000000000n,
      },
      "id": {
        "Concrete": {
          "interior": {
            "X2": [
              {
                "Parachain": 100,
              },
              {
                "GeneralKey": "USDT",
              },
            ],
          },
          "parents": 1,
        },
      },
    },
  },
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
              "network": "Any",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  1000000000,
]
`;

exports[`xTokens transferMultiCurrencies should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMulticurrencies",
  "getArgs": [Function],
  "module": "xTokens",
  "type": "Substrate",
}
`;

exports[`xTokens transferMultiCurrencies should get correct arguments 1`] = `
[
  [
    [
      "USDT",
      99000000000n,
    ],
    [
      "RMRK",
      5000000000n,
    ],
  ],
  1,
  {
    "V3": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;
